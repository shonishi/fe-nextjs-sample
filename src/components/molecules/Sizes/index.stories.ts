import type { Meta, StoryObj } from '@storybook/react';
import Sizes from '.';

const meta: Meta<typeof Sizes> = {
  component: Sizes,
};

export default meta;
type Story = StoryObj<typeof Sizes>;

export const MultipleElements: Story = {
  args: {
    sizes: [
      { id: 1, name: 'XXS', inStock: false },
      { id: 2, name: 'XS', inStock: true },
      { id: 3, name: 'S', inStock: true },
      { id: 4, name: 'M', inStock: true },
      { id: 5, name: 'L', inStock: true },
      { id: 6, name: 'XL', inStock: true },
      { id: 7, name: '2XL', inStock: true },
      { id: 8, name: '3XL', inStock: true },
    ],
    selectedSizeId: 3,
    changeSize: undefined,
  },
};

export const SingleElements: Story = {
  args: {
    sizes: [{ id: 2, name: 'XS', inStock: true }],
    selectedSizeId: 2,
    changeSize: undefined,
  },
};

export const EmptyElements: Story = {
  args: {
    sizes: [],
    selectedSizeId: 0,
    changeSize: undefined,
  },
};
