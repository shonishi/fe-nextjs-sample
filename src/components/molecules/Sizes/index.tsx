import { useAppDispatch } from '@/src/ducks/hooks';
import { RadioGroup } from '@headlessui/react';
import { ActionCreatorWithPayload } from '@reduxjs/toolkit';

interface Props {
  sizes: {
    id: number;
    name: string;
    inStock: boolean;
  }[];
  selectedSizeId: number;
  changeSize: ActionCreatorWithPayload<number, string>;
}

export default function Sizes({ sizes, selectedSizeId, changeSize }: Props) {
  const dispatch = useAppDispatch();
  const selectedSize = sizes.find((size) => size.id === selectedSizeId);
  return (
    <div className="mt-10">
      <div className="flex items-center justify-between">
        <h3 className="text-sm font-medium text-gray-900">Size</h3>
        <a
          href="#"
          className="text-sm font-medium text-indigo-600 hover:text-indigo-500"
        >
          Size guide
        </a>
      </div>

      <RadioGroup
        value={selectedSize}
        onChange={(e) => dispatch(changeSize(e.id))}
        className="mt-4"
      >
        <RadioGroup.Label className="sr-only"> Choose a size </RadioGroup.Label>
        <div className="grid grid-cols-4 gap-4 sm:grid-cols-8 lg:grid-cols-4">
          {sizes.map((size) => (
            <RadioGroup.Option
              key={size.name}
              value={size}
              disabled={!size.inStock}
              className={({ active }) =>
                [
                  size.inStock
                    ? 'cursor-pointer bg-white text-gray-900 shadow-sm'
                    : 'cursor-not-allowed bg-gray-50 text-gray-200',
                  active ? 'ring-2 ring-indigo-500' : '',
                  'group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6',
                ]
                  .filter(Boolean)
                  .join(' ')
              }
            >
              {({ active, checked }) => (
                <>
                  <RadioGroup.Label as="span">{size.name}</RadioGroup.Label>
                  {size.inStock ? (
                    <span
                      className={[
                        active ? 'border' : 'border-2',
                        checked ? 'border-indigo-500' : 'border-transparent',
                        'pointer-events-none absolute -inset-px rounded-md',
                      ].join(' ')}
                      aria-hidden="true"
                    />
                  ) : (
                    <span
                      aria-hidden="true"
                      className="pointer-events-none absolute -inset-px rounded-md border-2 border-gray-200"
                    >
                      <svg
                        className="absolute inset-0 h-full w-full stroke-2 text-gray-200"
                        viewBox="0 0 100 100"
                        preserveAspectRatio="none"
                        stroke="currentColor"
                      >
                        <line
                          x1={0}
                          y1={100}
                          x2={100}
                          y2={0}
                          vectorEffect="non-scaling-stroke"
                        />
                      </svg>
                    </span>
                  )}
                </>
              )}
            </RadioGroup.Option>
          ))}
        </div>
      </RadioGroup>
    </div>
  );
}
