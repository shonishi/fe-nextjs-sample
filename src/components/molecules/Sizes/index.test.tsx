import Sizes from '.';
import renderer from 'react-test-renderer';

describe('Sizes', () => {
  it('sizesの要素が複数の場合、全て表示されること', async () => {
    const tree = renderer
      .create(
        <Sizes
          {...{
            sizes: [
              { id: 1, name: 'XXS', inStock: false },
              { id: 2, name: 'XS', inStock: true },
              { id: 3, name: 'S', inStock: true },
              { id: 4, name: 'M', inStock: true },
              { id: 5, name: 'L', inStock: true },
              { id: 6, name: 'XL', inStock: true },
              { id: 7, name: '2XL', inStock: true },
              { id: 8, name: '3XL', inStock: true },
            ],
            selectedSizeId: 3,
            changeSize: jest.fn(),
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('sizesの要素が1つの場合、1つだけ表示されること', async () => {
    const tree = renderer
      .create(
        <Sizes
          {...{
            sizes: [{ id: 2, name: 'XS', inStock: true }],
            selectedSizeId: 2,
            changeSize: jest.fn(),
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('sizesが空の場合、何も表示されないこと', async () => {
    const tree = renderer
      .create(
        <Sizes
          {...{
            sizes: [],
            selectedSizeId: 0,
            changeSize: jest.fn(),
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
