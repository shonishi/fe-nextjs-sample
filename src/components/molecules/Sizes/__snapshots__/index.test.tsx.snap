// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sizes sizesが空の場合、何も表示されないこと 1`] = `
<div
  className="mt-10"
>
  <div
    className="flex items-center justify-between"
  >
    <h3
      className="text-sm font-medium text-gray-900"
    >
      Size
    </h3>
    <a
      className="text-sm font-medium text-indigo-600 hover:text-indigo-500"
      href="#"
    >
      Size guide
    </a>
  </div>
  <div
    aria-labelledby="headlessui-label-:rn:"
    className="mt-4"
    id="headlessui-radiogroup-:rm:"
    onKeyDown={[Function]}
    role="radiogroup"
  >
    <label
      className="sr-only"
      id="headlessui-label-:rn:"
    >
       Choose a size 
    </label>
    <div
      className="grid grid-cols-4 gap-4 sm:grid-cols-8 lg:grid-cols-4"
    />
  </div>
</div>
`;

exports[`Sizes sizesの要素が1つの場合、1つだけ表示されること 1`] = `
<div
  className="mt-10"
>
  <div
    className="flex items-center justify-between"
  >
    <h3
      className="text-sm font-medium text-gray-900"
    >
      Size
    </h3>
    <a
      className="text-sm font-medium text-indigo-600 hover:text-indigo-500"
      href="#"
    >
      Size guide
    </a>
  </div>
  <div
    aria-labelledby="headlessui-label-:rj:"
    className="mt-4"
    id="headlessui-radiogroup-:ri:"
    onKeyDown={[Function]}
    role="radiogroup"
  >
    <label
      className="sr-only"
      id="headlessui-label-:rj:"
    >
       Choose a size 
    </label>
    <div
      className="grid grid-cols-4 gap-4 sm:grid-cols-8 lg:grid-cols-4"
    >
      <div
        aria-checked="true"
        aria-labelledby="headlessui-label-:rl:"
        className="cursor-pointer bg-white text-gray-900 shadow-sm group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6"
        data-headlessui-state="checked"
        id="headlessui-radiogroup-option-:rk:"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        role="radio"
        tabIndex={0}
      >
        <span
          id="headlessui-label-:rl:"
        >
          XS
        </span>
        <span
          aria-hidden="true"
          className="border-2 border-indigo-500 pointer-events-none absolute -inset-px rounded-md"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Sizes sizesの要素が複数の場合、全て表示されること 1`] = `
<div
  className="mt-10"
>
  <div
    className="flex items-center justify-between"
  >
    <h3
      className="text-sm font-medium text-gray-900"
    >
      Size
    </h3>
    <a
      className="text-sm font-medium text-indigo-600 hover:text-indigo-500"
      href="#"
    >
      Size guide
    </a>
  </div>
  <div
    aria-labelledby="headlessui-label-:r1:"
    className="mt-4"
    id="headlessui-radiogroup-:r0:"
    onKeyDown={[Function]}
    role="radiogroup"
  >
    <label
      className="sr-only"
      id="headlessui-label-:r1:"
    >
       Choose a size 
    </label>
    <div
      className="grid grid-cols-4 gap-4 sm:grid-cols-8 lg:grid-cols-4"
    >
      <div
        aria-checked="false"
        aria-disabled={true}
        aria-labelledby="headlessui-label-:r3:"
        className="cursor-not-allowed bg-gray-50 text-gray-200 group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6"
        data-headlessui-state="disabled"
        id="headlessui-radiogroup-option-:r2:"
        role="radio"
        tabIndex={-1}
      >
        <span
          id="headlessui-label-:r3:"
        >
          XXS
        </span>
        <span
          aria-hidden="true"
          className="pointer-events-none absolute -inset-px rounded-md border-2 border-gray-200"
        >
          <svg
            className="absolute inset-0 h-full w-full stroke-2 text-gray-200"
            preserveAspectRatio="none"
            stroke="currentColor"
            viewBox="0 0 100 100"
          >
            <line
              vectorEffect="non-scaling-stroke"
              x1={0}
              x2={100}
              y1={100}
              y2={0}
            />
          </svg>
        </span>
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-:r5:"
        className="cursor-pointer bg-white text-gray-900 shadow-sm group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6"
        data-headlessui-state=""
        id="headlessui-radiogroup-option-:r4:"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        role="radio"
        tabIndex={-1}
      >
        <span
          id="headlessui-label-:r5:"
        >
          XS
        </span>
        <span
          aria-hidden="true"
          className="border-2 border-transparent pointer-events-none absolute -inset-px rounded-md"
        />
      </div>
      <div
        aria-checked="true"
        aria-labelledby="headlessui-label-:r7:"
        className="cursor-pointer bg-white text-gray-900 shadow-sm group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6"
        data-headlessui-state="checked"
        id="headlessui-radiogroup-option-:r6:"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        role="radio"
        tabIndex={0}
      >
        <span
          id="headlessui-label-:r7:"
        >
          S
        </span>
        <span
          aria-hidden="true"
          className="border-2 border-indigo-500 pointer-events-none absolute -inset-px rounded-md"
        />
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-:r9:"
        className="cursor-pointer bg-white text-gray-900 shadow-sm group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6"
        data-headlessui-state=""
        id="headlessui-radiogroup-option-:r8:"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        role="radio"
        tabIndex={-1}
      >
        <span
          id="headlessui-label-:r9:"
        >
          M
        </span>
        <span
          aria-hidden="true"
          className="border-2 border-transparent pointer-events-none absolute -inset-px rounded-md"
        />
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-:rb:"
        className="cursor-pointer bg-white text-gray-900 shadow-sm group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6"
        data-headlessui-state=""
        id="headlessui-radiogroup-option-:ra:"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        role="radio"
        tabIndex={-1}
      >
        <span
          id="headlessui-label-:rb:"
        >
          L
        </span>
        <span
          aria-hidden="true"
          className="border-2 border-transparent pointer-events-none absolute -inset-px rounded-md"
        />
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-:rd:"
        className="cursor-pointer bg-white text-gray-900 shadow-sm group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6"
        data-headlessui-state=""
        id="headlessui-radiogroup-option-:rc:"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        role="radio"
        tabIndex={-1}
      >
        <span
          id="headlessui-label-:rd:"
        >
          XL
        </span>
        <span
          aria-hidden="true"
          className="border-2 border-transparent pointer-events-none absolute -inset-px rounded-md"
        />
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-:rf:"
        className="cursor-pointer bg-white text-gray-900 shadow-sm group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6"
        data-headlessui-state=""
        id="headlessui-radiogroup-option-:re:"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        role="radio"
        tabIndex={-1}
      >
        <span
          id="headlessui-label-:rf:"
        >
          2XL
        </span>
        <span
          aria-hidden="true"
          className="border-2 border-transparent pointer-events-none absolute -inset-px rounded-md"
        />
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-:rh:"
        className="cursor-pointer bg-white text-gray-900 shadow-sm group relative flex items-center justify-center rounded-md border py-3 px-4 text-sm font-medium uppercase hover:bg-gray-50 focus:outline-none sm:flex-1 sm:py-6"
        data-headlessui-state=""
        id="headlessui-radiogroup-option-:rg:"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        role="radio"
        tabIndex={-1}
      >
        <span
          id="headlessui-label-:rh:"
        >
          3XL
        </span>
        <span
          aria-hidden="true"
          className="border-2 border-transparent pointer-events-none absolute -inset-px rounded-md"
        />
      </div>
    </div>
  </div>
</div>
`;
